### PR TITLE
Add support for writing specs as HRX files

### PR DESCRIPTION
**Pull request by nex3**
See https://github.com/google/hrx

### CODE REVIEW

**Comment by nschonni** [comment]
Is the plan for people to write parsers for other languages for HRX or is that something you're working on with your team?

**Comment by glebm** [comment]
Is it worth it?

You lose:

1. Syntax highlighting on GitHub and in the IDE.
2. Compatibility with standard unix tools like `find` and `grep` (they can longer filter just inputs or just outputs).
3. Some git history usefulness.

**Comment by nex3** [comment]
> Is the plan for people to write parsers for other languages for HRX or is that something you're working on with your team?

My "team" at the moment is really only me a couple part-time volunteers whose hands are full :stuck_out_tongue_closed_eyes:. I think in general, if there are consumers of the specs outside of this repo, it's up to those consumers to stay up-to-date with whichever spec format works best for the velocity and maintainability of this repo and its role in language development and verification. However, there are already third-party implementations in [Node.js](https://www.npmjs.com/package/node-hrx) and [Python](https://github.com/rebeccajae/hrx.py).

> Is it worth it?

I strongly believe this is worth it. The current system of having many files for each test case causes many severe usability problems. It's currently *very* difficult, in local editing and especially in GitHub's user interface, to understand everything about a given test case at once. Adding new test cases creates colossal pull requests full of a sea of undifferentiated tiny files; editing an existing test case often creates a pull request that only shows either its input or its output, but not both; error tests create giant directories that can't be manually ordered by human authors; and so on.

As an example, compare how difficult it is to understand exactly what's being tested in [the specs for `rgb()` and `rgba()`](https://github.com/sass/sass-spec/tree/master/spec/core_functions/color) versus [the original Ruby Sass test cases](https://github.com/sass/ruby-sass/blob/stable/test/sass/functions_test.rb#L209-L283). Even worse, imagine you're a code reviewer looking at [the pull request that added those specs](https://github.com/sass/sass-spec/pull/1300/files) and see how long it takes you to figure out whether I missed an edge-case error test.

HRX won't get us to the degree of terseness that writing tests in code can, but it will get us a lot closer to their locality and reviewability.

> You lose:
> 
> 1. Syntax highlighting on GitHub and in the IDE.

This is true, but specs are usually so simple that standard syntax highlighting is largely irrelevant. HRX is also a very simple format that would be pretty easy to add syntax highlighting for to any system (as long as it supports calling out to highlighters for other languages).

> 2. Compatibility with standard unix tools like `find` and `grep` (they can longer filter just inputs or just outputs).

I personally only ever use `grep` across all the files in the repo, inputs *or* outputs. If something shows up in an input file, it almost always shows up only in outputs that correspond to that input anyway, or it's a Sass-specific construct that doesn't appear in outputs at all. Do you have an example where it's important to search only one or the other?

> 3. Some git history usefulness.

This is true for any refactor. We've weathered it before with the change to nested syntax by default, and we'll weather it again when #1074 gets fixed.

**Comment by nschonni** [comment]
I'm not going to 👎, but I'm also not 👍 on this. I think a MD/asciidoc approach is better in a tooling and publishing perspective than a new format. GitHub and Jekyll supports converting those formats into a published spec easier. Not that the approach wouldn't require docs and convention as well. It would also break the grep flow @glebm mentioned.

**Comment by nex3** [comment]
> I'm not going to , but I'm also not  on this. I think a MD/asciidoc approach is better in a tooling and publishing perspective than a new format.

I understand that a Markdown (or asciidoc) format may be more readable when browsing specs on GitHub, at least until HRX gets syntax highlighting support. But it seems substantially *less* toolable. Standard libraries for those formats are generally not set up to make it easy to extract structured data from them, and *certainly* not to update that data in a way that otherwise preserves the formatting of the files. There's not even a standard way of defining "these files contain this text"; we'd have to make an *ad hoc* format on top of an existing text format. HRX, on the other hand, is well-specified and comes with libraries that can read (Ruby, JS, Python) and write (Ruby) files in a standard way out of the box.

> GitHub and Jekyll supports converting those formats into a published spec easier. Not that the approach wouldn't require docs and convention as well.

I think the idea of combining the language's test suite with its prose specification, while appealing on the surface, is ultimately unworkable. The [prose specifications we have now](https://github.com/sass/language/tree/master/spec) work fine without being tied to executable test cases. On the other hand, when we write test cases, they should be very thorough--which is not something anyone wants to slog through when trying to read a spec.

It's definitely valuable to have some amount of explanatory text in our test cases. But inline comments, supplemented with the occasional README file, have been sufficient for that so far and I expect they'll continue to be.

> It would also break the grep flow @glebm mentioned.

I still don't understand the specific use-cases this breaks.

**Comment by glebm** [comment]
> I still don't understand the specific use-cases this breaks.

I often write ad-hoc bash scripts to make changes, such as this one: https://github.com/sass/sass-spec/pull/1319

This ad-hoc stuff will have to be done with Ruby. Not a problem for me, as I'm comfortable with Ruby, but it will probably break some workflow.

Another couple of examples, from the "rsass" work-in-progress Rust Sass implementation:
1. sass-spec runner, Bash: https://github.com/kaj/rsass/blob/master/check_spec
2. sass-spec -> rust test converter, Rust (this will need a rust hrx reader library): https://github.com/kaj/rsass/blob/master/src/spectest.rs#L171

**Comment by nex3** [comment]
> > I still don't understand the specific use-cases this breaks.
> 
> I often write ad-hoc bash scripts to make changes, such as this one: #1319
>
> This ad-hoc stuff will have to be done with Ruby. Not a problem for me, as I'm comfortable with Ruby, but it will probably break some workflow.

Isn't this sort of thing that the spec runner's annotate, generate, and interactive modes are for? For your example, you could write:

```bash
rm spec/output_styles/options.yml
sass-spec.rb --generate ${NESTED[@]} ${COMPACT[@]} ${COMPRESSED[@]}
git checkout spec/output_styles/options.yml
```

You can probably even avoid modifying `spec/output_styles/options.yml` itself using `sass-spec.rb annotate`.

> Another couple of examples, from the "rsass" work-in-progress Rust Sass implementation:
> 
> 1. sass-spec runner, Bash: https://github.com/kaj/rsass/blob/master/check_spec

That also seems like the sort of thing that annotate mode would handle well.

Also, it's worth noting that both this and the previous example are pretty specific to the output style specs, which will be removed with https://github.com/sass/sass-spec/issues/1305.

> 2. sass-spec -> rust test converter, Rust (this will need a rust hrx reader library): https://github.com/kaj/rsass/blob/master/src/spectest.rs#L171

Yes, this sort of thing will need to be updated, but that's no less true for any format we choose. HRX is very simple and clearly specified, so it will almost certainly be easier to implement clients like this for than an *ad hoc* Markdown-based format.

**Comment by glebm** [comment]
> Isn't this sort of thing that the spec runner's annotate, generate, and interactive modes are for? 

Thanks, I didn't know that the spec runner can do this!

**Comment by nshahan** [comment]
Currently the cognitive load of reviewing changes to the spec is huge. When adding a new features to the language reviewing the changes can become a major hurdle. Here is another recent example: https://github.com/sass/sass-spec/pull/1303/files

If the Sass language were in maintenance mode I'd probably argue for keeping the existing spec runner and inertia behind it, but that is not the case. There are multiple large proposals for new features coming. I feel that having a sane way for humans to review changes to the spec would likely help ensure those new features land cleanly.

**Comment by nex3** [comment]
I'm getting to the point where I want to start landing specs for the module system, and without this feature the authoring/review burden will be huge. Can we compromise on landing this while also making sure `annotate.rb` *et al* are up-to-snuff for use-cases like @glebm's?

**Comment by nschonni** [comment]
@xzyfer I don't plan on writing a library to support this format for node-sass. Any objections to dropping our spec testing after this lands?

**Comment by nex3** [comment]
@nschonni There's already [a node library for HRX](https://www.npmjs.com/package/node-hrx), so I think all you'd need to do to keep the Node Sass tests running is update the runner to use that library.

**Comment by nschonni** [comment]
@nex3 fair enough, I probably just have some unrelated biases from interviewing with SF companies, but to me this feels like a "not invented here" thing https://xkcd.com/927/ 
Maybe someone else will be interested in implementing. I'll leave it up to @xzyfer if it gets pulled (sorry to pass the buck)

**Comment by nschonni** [comment]
node-sass testing with the NPM published versions of the spec, so you don't need to worry about breaking anything by landing this.

**Comment by nex3** [comment]
That's great! I'd still really like to get an official GitHub approval before I land, though.

**Comment by nex3** [comment]
Re-assigning reviewers. I need to land this so I can start working on module system specs. If anyone has any remaining concerns about use-cases that aren't covered by the existing repo scripts, please file an issue and assign it to me.

File: lib/sass_spec/directory.rb [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+          _write!
+        end
+      end
+    else
+      FileUtils.rm_rf(@path)
+    end
+  end
+
+  # If this directory refers to an HRX file, runs a block the that archive's
+  # directory and all its contents physically present on the filesystem.
```
[the remark this answers is not in the document]
Comment (nex3) [reply lib/sass_spec/directory.rb]: Done.
